src/service/locator.cpp

```diff
@@ -205,121 +196,141 @@ locator_t::connect_sink_t::on_shutdown() {
 }
 
 class locator_t::publish_slot_t: public basic_slot<locator::publish> {
-    struct publish_lock_t: public basic_slot<locator::publish>::dispatch_type {
+    locator_t *const parent;
+
+    class publish_sink_t: public basic_slot::dispatch_type {
         publish_slot_t *const parent;
         std::string     const handle;
 
-        publish_lock_t(publish_slot_t * const parent_, const std::string& handle_):
-            basic_slot<locator::publish>::dispatch_type("publish"),
+    public:
+        publish_sink_t(publish_slot_t *const parent_, const std::string& handle_):
+            basic_slot::dispatch_type("publish/discard"),
             parent(parent_),
             handle(handle_)
         {
-            on<locator::publish::discard>([this] { discard({}); });
+            on<locator::publish::discard>([this] { discard(std::error_code{}); });
         }
 
         virtual
         void
-        discard(const std::error_code& ec) const { parent->discard(ec, handle); }
+        discard(const std::error_code& ec) const { parent->on_discard(ec, handle); }
     };
 
-    typedef std::shared_ptr<const basic_slot::dispatch_type> result_type;
-
-    locator_t *const parent;
+    // Slot's transition type.
+    typedef boost::optional<std::shared_ptr<const basic_slot::dispatch_type>> result_type;
 
 public:
     publish_slot_t(locator_t *const parent_): parent(parent_) { }
 
     auto
-    operator()(tuple_type&& args, upstream_type&& upstream) -> boost::optional<result_type> {
-        const auto dispatch = cocaine::tuple::invoke(std::move(args),
-            [this](std::string&& handle, std::vector<tcp::endpoint>&& location,
-                   std::tuple<unsigned int, graph_root_t>&& metadata) -> result_type
-        {
-            scoped_attributes_t attributes(*parent->m_log, { attribute::make("service", handle) });
+    operator()(tuple_type&& args, upstream_type&& upstream) -> result_type;
 
-            unsigned int versions;
-            graph_root_t protocol;
+private:
+    void
+    on_discard(const std::error_code& ec, const std::string& handle);
+};
 
-            std::tie(versions, protocol) = metadata;
+auto
+locator_t::publish_slot_t::operator()(tuple_type&& args, upstream_type&& upstream) -> result_type {
+    const auto dispatch = cocaine::tuple::invoke(std::move(args),
+        [this](std::string&& handle, std::vector<tcp::endpoint>&& location,
```

Comment:
Forced rvalue references here looks... strange. We can just pass by value, aren't we?